### ISSUE TITLE
Bug: X-Ray sampling configuration settings are ignored when using AWS ALB in front of App Mesh

### ISSUE THREAD

**Issue by cageyv** [aws/aws-app-mesh-roadmap#387]
*SECURITY NOTICE: If you think you’ve found a potential security issue, please do not post it in the Issues. Instead, please follow the instructions [here](https://aws.amazon.com/security/vulnerability-reporting/) or [email AWS security directly](mailto:aws-security@amazon.com).*

**Summary**
If we use ALB in front of the service with sidecar AppMesh/Envoy container and want to enable the X-Ray tracing function, all 100% traces will be recorded and the tracing settings on the Envoy side will be ignored

**Steps to Reproduce**
- Create ALB with Target Group
- Create ECS Service with AppMesh Envoy sidecar and X-Ray Deamon sidecar and attach to ALB/Target group
- Enable X-Ray tracing on AppMesh Envoy `ENABLE_ENVOY_XRAY_TRACING = "1", XRAY_DAEMON_PORT = "2000"`
- Set the X-Ray Sampling rate to 0.01% `XRAY_SAMPLING_RATE = "0.01"` or set custom manifest and disable sampling at all `XRAY_SAMPLING_RULE_MANIFEST = "/tmp/sampling.json"`
```json
{
    "version": 2,
    "rules": [
        {
            "description": "NoEnvoyHealthChecks",
            "host": "*",
            "http_method": "*",
            "url_path": "/ping",
            "fixed_target": 0,
            "rate": 0
        }
    ],
    "default": {
        "fixed_target": 0,
        "rate": 0
    }
}
```
- Set Envoy logs to trace level `ENVOY_LOG_LEVEL = "trace"` 
- (Optional) Add logging request headers inside your application 
- Send multiple requests to the service so that they go through the ALB
- Check the result in the AWS X-Ray console. Each request will be traced. 
- Check the value of the metrics in CloudWatch. ApproximateTraceCount for X-Ray and RequestCount for Target group. Set statistic to Sum. If there are no other X-Ray traces, then these two plots will completely match.
- Dig into Envoy logs and search for `x-amzn-trace-id` value. It will contain `Sampled=1`. 
Example: `'x-amzn-trace-id', 'Root=1-61eed92c-d9e13998920742ff89cd8525;Parent=539a369bae06888c;Sampled=1'` 

**Are you currently working around this issue?**
A temporary solution could be to integrate the X-Ray SDK inside the application and disable tracing on Envoy

**Additional context**
- ALB adds `Sampled=0` to header `x-amzn-trace-id` when sending health check endpoint requests
- There is no way to disable adding `x-amzn-trace-id` header on ALB side
- We can't see the full value of `x-amzn-trace-id` header inside ALB Access logs. Probably because of this limitation
`If the HTTP headers are greater than 7 KB, the load balancer rewrites the X-Amzn-Trace-Id header with a Root field.`

**Attachments**
![cw_trace_count_1](https://user-images.githubusercontent.com/51059484/151136016-e9ff1335-b687-4bdc-824f-e345261e21d0.jpg)

**Comment by suniltheta** [comment]
Thank you very much for filing the issue and the detailed explanation. This is an important discovery where Envoy's XRay tracer extension samples 100% of the requests originating from ALB.

The only way to mitigate this issue for now (until a new AppMesh envoy image is released) is to disable xray tracing on the envoy receiving request directly from ALB.

The next release of App Mesh envoy image should contain the fix for this issue.

**Comment by suniltheta** [comment]
This issue is fixed and released in v1.21.0.0 https://github.com/aws/aws-app-mesh-roadmap/issues/389.
Closing this issue
